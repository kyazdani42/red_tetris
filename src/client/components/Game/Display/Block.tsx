import * as React from 'react';

export const NextBlock = (color: string, key: string) => (
  <div
    style={{
      backgroundColor: changeBlockColor(color),
      border: '1px solid black',
      width: 'calc(25% - 2px)',
      height: 'calc(25% - 2px)'
    }}
    key={key}
  />
);

export const Block = (color: string, key: string) => (
  <div
    style={{
      backgroundColor: changeBlockColor(color),
      border: '1px solid black',
      width: 'calc(10% - 2px)',
      height: 'calc(5% - 2px)'
    }}
    key={key}
  />
);

export const changeBlockColor = (color: string) => {
  switch (color) {
    case 'skyBlue':
     return '#adf0ff';
    case 'blue':
      return '#7427CC';
    case 'orange':
      return '#EF6A00';
    case 'yellow':
      return '#FFC851';
    case 'green':
      return '#EFA300';
    case 'purple':
      return '#650000';
    case 'red':
      return '#B11515';
    default:
      return color;
  }
};
