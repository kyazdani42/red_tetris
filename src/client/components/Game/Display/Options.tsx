import * as React from 'react';
import { connect } from 'react-redux';

import { setOptions } from '../../../actions/actions';
import { State } from '../../../store';
import { OptionLabelWrapper, OptionsWrapper } from './styles';

interface Props {
  options: Options;
  dispatchSetOptions: (options: Options) => void;
}

const handleDispatch = (params: Props) => (option: { name: string; on: boolean }) =>
  params.dispatchSetOptions({ ...params.options, [option.name]: option.on });

export const Options: React.SFC<Props> = ({ options, dispatchSetOptions }) => {
  const dispatch = handleDispatch({ options, dispatchSetOptions });
  return (
    <OptionsWrapper>
      {checkbox(options.invisible, 'invisible', dispatch)}
      {checkbox(options.mirror, 'mirror', dispatch)}
      {checkbox(options.reverse, 'reverse', dispatch)}
    </OptionsWrapper>
  );
};

const checkbox = (
  checked: boolean,
  name: string,
  dispatch: (option: { name: string; on: boolean }) => void
) => (
  <OptionLabelWrapper>
    <label>{name}</label>
    <input type="checkbox" checked={checked} onChange={() => dispatch({ name, on: !checked })} />
  </OptionLabelWrapper>
);

const mapStateToProps = (state: State) => ({
  options: state.app.options
});

const mapDispatchToProps = (dispatch: any) => ({
  dispatchSetOptions: (options: Options) => dispatch(setOptions(options))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Options);
