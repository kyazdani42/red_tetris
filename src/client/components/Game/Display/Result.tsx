import * as React from 'react';

import { ResultStyle } from './styles';

interface Props {
  winner: boolean;
}
export const Result: React.SFC<Props> = ({ winner }) => (
  <ResultStyle winner={winner}>{winner ? 'You Win!' : 'Game Over!'}</ResultStyle>
);
