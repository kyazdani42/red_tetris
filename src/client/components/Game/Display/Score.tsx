import * as React from 'react';

import { ScoreStyle } from './styles';

interface Props {
  score: number;
}

export const Score: React.SFC<Props> = ({ score }) => (
  <ScoreStyle>
    score: {score}
  </ScoreStyle>
);
