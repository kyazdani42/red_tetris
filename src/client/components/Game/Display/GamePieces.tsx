import * as React from 'react';
import styled from 'styled-components';

import { Block } from './Block';
import { GamePiecesContainer } from './styles';

interface Props {
  stack: GameProps['stack'];
}

const getDivsFromStack = (stack: GameProps['stack']) => {
  const container = Array(200);
  for (let i = 0; i < 20; i++) {
    for (let j = 0; j < 10; j++) {
      const containerIndex = +`${i}${j}`;
      container[containerIndex] = (
        <Block color={stack[i][j].color} key={'block' + containerIndex} />
      );
    }
  }
  return container;
};

export const GamePieces: React.SFC<Props> = ({ stack }) => {
  if (!stack.length) return null;
  return <GamePiecesContainer>{getDivsFromStack(stack)}</GamePiecesContainer>;
};
