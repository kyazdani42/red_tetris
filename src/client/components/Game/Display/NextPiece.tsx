import * as React from 'react';

import { NextBlock } from './Block';
import { PieceContainer } from './styles';

interface Props {
  nextPiece: {
    color: string;
    patterns: number[][][];
  } | null;
}

export const NextPiece: React.SFC<Props> = ({ nextPiece }) => (
  <PieceContainer>
    {getBlocksFromPiece(nextPiece)}
  </PieceContainer>
);

const getBlocksFromPiece = (nextPiece: Props['nextPiece']) => {
  if (!nextPiece) return null;
  const piece = [];
  const len = nextPiece.patterns[0].length;
  for (let i = 0; i < 4; i++) {
    const line = i < len ? nextPiece.patterns[0][i] : null;
    for (let j = 0; j < 4; j++) {
      const blockValue = line ? j < line.length ? line[j] : 0 : 0;
      const key = `block${i}-${j}-next`;
      if (blockValue) {
        piece.push(<NextBlock color={nextPiece.color} key={key} />);
      } else {
        piece.push(<NextBlock color="black" key={key} />);
      }
    }
  }
  return piece;
};
