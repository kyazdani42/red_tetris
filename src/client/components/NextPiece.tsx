import * as React from 'react';
import styled from 'styled-components';

import { NextBlock } from './Block';

interface Props {
  nextPiece: {
    color: string;
    patterns: number[][][];
  } | null;
}

const PieceContainer = styled.div`
  position: absolute;
  top: 41px;
  left: -173px;
  height: 170px;
  width: 170px;
  border: 1px solid #fff;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  background-color: #000;
`;

export const NextPiece: React.SFC<Props> = ({ nextPiece }) => (
  <PieceContainer>
    {getBlocksFromPiece(nextPiece)}
  </PieceContainer>
);

const getBlocksFromPiece = (nextPiece: Props['nextPiece']) => {
  if (!nextPiece) return null;
  const piece = [];
  const len = nextPiece.patterns[0].length;
  for (let i = 0; i < 4; i++) {
    const line = i < len ? nextPiece.patterns[0][i] : null;
    for (let j = 0; j < 4; j++) {
      const blockValue = line ? j < line.length ? line[j] : 0 : 0;
      const key = `block${i}-${j}-next`;
      if (blockValue) {
        piece.push(<NextBlock color={nextPiece.color} key={key} />);
      } else {
        piece.push(<NextBlock color="black" key={key} />);
      }
    }
  }
  return piece;
};
