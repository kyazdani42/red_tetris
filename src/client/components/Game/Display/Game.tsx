import * as React from 'react';
import { connect } from 'react-redux';

import { State } from '../../../redux/store';
import { GamePieces } from './GamePieces';
import { NextPiece } from './NextPiece';
import Options from './Options';
import { Result } from './Result';
import { Score } from './Score';
import StartButton from './StartButton';
import { GameStartInfoWrapper } from './styles';

interface Props {
  gameData: GameProps | null;
  socket: SocketIOClient.Socket;
}

export const Game: React.SFC<Props> = ({ gameData, socket }) => {
  let gamePieces, score, nextPiece, startButton, result, options;
  const isPlaying = gameData && gameData.isPlaying || false;
  if (gameData) {
    score = gameData.score;
    nextPiece = gameData.nextPiece;

    const { stack, running, isOwner, winner } = gameData;
    gamePieces = stack.length ? <GamePieces stack={stack} /> : null;
    startButton = running || !isOwner ? null : <StartButton socket={socket} />;
    options = running || !isOwner ? null : <Options />;
    result = winner === undefined || isPlaying ? null : <Result winner={winner} />;
  } else {
    score = 0;
    nextPiece = null;
    gamePieces = null;
    startButton = null;
    result = null;
    options = null;
  }
  return (
    <React.Fragment>
      {gamePieces}
      <Score score={score} />
      <NextPiece nextPiece={nextPiece} />
      {!isPlaying &&
      <GameStartInfoWrapper>
        {startButton}
        {options}
        {result}
      </GameStartInfoWrapper>
      }
    </React.Fragment>
  );
};

const mapStateToProps = ({ app }: State) => ({
  gameData: app.gameData,
  socket: app.socket
});

export default connect(mapStateToProps)(Game);
