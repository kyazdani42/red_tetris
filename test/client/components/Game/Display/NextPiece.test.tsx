// tslint:disable
import * as React from 'react';
import { shallow } from 'enzyme';

import { NextPiece, getBlocksFromPiece } from '../../../../../src/client/components/Game/Display/NextPiece';

const getPiece = (color: string) => ({
  color,
  patterns: [
    [
      [1,1,0],[1,1,0],
    ],
  ],
});

describe('testing NextPiece component', () => {
  const wrapper = shallow(<NextPiece nextPiece={getPiece('blue')} />);
  it('smoke tests the component', () => {
    expect(wrapper.update()).toMatchSnapshot();
  })
})

describe('testing getBlocksFromPiece', () => {
  it('returns null if nextpiece is null', () => {
    expect(getBlocksFromPiece(null)).toBe(null);
  })
})