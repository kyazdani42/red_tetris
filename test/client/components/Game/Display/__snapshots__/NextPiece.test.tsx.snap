// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing NextPiece component smoke tests the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    nextPiece={
      Object {
        "color": "blue",
        "patterns": Array [
          Array [
            Array [
              1,
              1,
              0,
            ],
            Array [
              1,
              1,
              0,
            ],
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Unknown
          color="blue"
        />,
        <Unknown
          color="blue"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="blue"
        />,
        <Unknown
          color="blue"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
        <Unknown
          color="black"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "block0-0-next",
        "nodeType": "function",
        "props": Object {
          "color": "blue",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block0-1-next",
        "nodeType": "function",
        "props": Object {
          "color": "blue",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block0-2-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block0-3-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block1-0-next",
        "nodeType": "function",
        "props": Object {
          "color": "blue",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block1-1-next",
        "nodeType": "function",
        "props": Object {
          "color": "blue",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block1-2-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block1-3-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block2-0-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block2-1-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block2-2-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block2-3-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block3-0-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block3-1-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block3-2-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "block3-3-next",
        "nodeType": "function",
        "props": Object {
          "color": "black",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-htpNat",
        "isStatic": true,
        "rules": Array [
          "
  z-index: 10;
  position: absolute;
  bottom: -1px;
  left: calc(-17vmin - 3px);
  height: 17vmin;
  width: 17vmin;
  border: 1px solid #fff;
  border-right: none;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  background-color: #000;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-htpNat",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Unknown
            color="blue"
          />,
          <Unknown
            color="blue"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="blue"
          />,
          <Unknown
            color="blue"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
          <Unknown
            color="black"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "block0-0-next",
          "nodeType": "function",
          "props": Object {
            "color": "blue",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block0-1-next",
          "nodeType": "function",
          "props": Object {
            "color": "blue",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block0-2-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block0-3-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block1-0-next",
          "nodeType": "function",
          "props": Object {
            "color": "blue",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block1-1-next",
          "nodeType": "function",
          "props": Object {
            "color": "blue",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block1-2-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block1-3-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block2-0-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block2-1-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block2-2-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block2-3-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block3-0-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block3-1-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block3-2-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "block3-3-next",
          "nodeType": "function",
          "props": Object {
            "color": "black",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-htpNat",
          "isStatic": true,
          "rules": Array [
            "
  z-index: 10;
  position: absolute;
  bottom: -1px;
  left: calc(-17vmin - 3px);
  height: 17vmin;
  width: 17vmin;
  border: 1px solid #fff;
  border-right: none;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  background-color: #000;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-htpNat",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
